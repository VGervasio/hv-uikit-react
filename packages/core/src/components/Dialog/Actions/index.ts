import { getClasses } from "utils";

export type HvDialogActionClasses = {
  root?: string;
  fullscreen?: string;
  spacing?: string;
};

const classKeys: string[] = ["root", "fullscreen", "spacing"];

export const dialogActionClasses = getClasses<HvDialogActionClasses>(
  classKeys,
  "HvDialog-Action"
);

export * from "./Actions";
