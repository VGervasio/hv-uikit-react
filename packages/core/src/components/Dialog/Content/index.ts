import { getClasses } from "utils";

export type HvDialogContentClasses = {
  root?: string;
};

const classKeys: string[] = ["root"];

export const dialogContentClasses = getClasses<HvDialogContentClasses>(
  classKeys,
  "HvDialog-Content"
);

export * from "./Content";
