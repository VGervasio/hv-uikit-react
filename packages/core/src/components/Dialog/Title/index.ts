export { default as dialogTitleClasses } from "./titleClasses";
export * from "./titleClasses";
export * from "./Title";
