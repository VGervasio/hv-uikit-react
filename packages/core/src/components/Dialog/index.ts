export * from "./Dialog";
export { HvTitle as HvDialogTitle } from "./Title";
export { HvContent as HvDialogContent } from "./Content";
export { HvActions as HvDialogActions } from "./Actions";
