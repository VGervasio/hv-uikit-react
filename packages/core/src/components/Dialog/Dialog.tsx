import React, { useCallback, useMemo } from "react";
import { ClassNames } from "@emotion/react";
import MuiDialog, { DialogProps as MuiDialogProps } from "@mui/material/Dialog";
import { BackdropProps } from "@mui/material";

import { Close } from "@hitachivantara/uikit-react-icons";
import { theme } from "@hitachivantara/uikit-styles";
import { HvBaseProps } from "@core/types/generic";
import { setId } from "@core/utils";
import { withTooltip } from "@core/hocs";
import { useTheme } from "@core/hooks";
import dialogClasses, { HvDialogClasses } from "./dialogClasses";
import { StyledBackdrop, StyledClose, styles } from "./Dialog.styles";

export interface HvDialogProps
  extends Omit<MuiDialogProps, "fullScreen" | "classes" | "open">,
    HvBaseProps {
  /** Id to be applied to the root node. */
  id?: string;
  /** Current state of the Dialog. */
  open?: boolean;
  /** Function executed on close. */
  onClose?: (
    event: React.SyntheticEvent,
    reason?: "escapeKeyDown" | "backdropClick"
  ) => void;
  /** @inheritdoc */
  maxWidth?: MuiDialogProps["maxWidth"];
  /** @inheritdoc */
  fullWidth?: MuiDialogProps["fullWidth"];
  /**
   * Element id that should be focus when the Dialog opens.
   * Auto-focusing elements can cause usability issues, so this should be avoided.
   * @deprecated Use `autoFocus` on the element instead, if auto-focusing is required.
   */
  firstFocusable?: string;
  /** Title for the button close. */
  buttonTitle?: string;
  /** Set the dialog to fullscreen mode. */
  fullscreen?: boolean;
  /** Prevent closing the dialog when clicking on the backdrop. */
  disableBackdropClick?: boolean;
  /** A Jss Object used to override or extend the styles applied to the component. */
  classes?: HvDialogClasses;
}

const DialogBackdrop = (backdropProps: BackdropProps) => {
  const { activeTheme, selectedMode } = useTheme();
  return (
    <StyledBackdrop
      $backColor={
        activeTheme?.colors?.modes[selectedMode].atmo4 || theme.colors.atmo4
      }
      {...backdropProps}
    />
  );
};

export const HvDialog = ({
  classes,
  className,
  id,
  children,
  open = false,
  onClose,
  firstFocusable,
  buttonTitle = "Close",
  fullscreen = false,
  disableBackdropClick = false,
  ...others
}: HvDialogProps) => {
  delete (others as any).fullScreen;

  const { rootId } = useTheme();

  // Because the `disableBackdropClick` property was deprecated in MUI5
  // and we want to maintain that functionality to the user we're wrapping
  // the onClose call here to make that check.
  const wrappedClose = useCallback(
    (
      event,
      bypassValidation: boolean = false,
      reason?: "escapeKeyDown" | "backdropClick"
    ) => {
      if (bypassValidation || !disableBackdropClick) {
        onClose?.(event, reason);
      }
    },
    [onClose]
  );

  const measuredRef = useCallback(() => {
    if (!firstFocusable) return;

    const element = document.getElementById(firstFocusable);
    element?.focus();
  }, [firstFocusable]);

  const closeButtonDisplay = () => <Close role="presentation" />;

  const CloseButtonTooltipWrapper = buttonTitle
    ? withTooltip(closeButtonDisplay, buttonTitle, "top")
    : closeButtonDisplay;

  const slots = useMemo<MuiDialogProps["slots"]>(
    () => ({
      backdrop: (backdropProps) => (
        <DialogBackdrop open={open} onClick={wrappedClose} {...backdropProps} />
      ),
    }),
    [open, wrappedClose]
  );

  return (
    <ClassNames>
      {({ css, cx }) => (
        <MuiDialog
          container={document.getElementById(rootId || "") || document.body}
          className={cx(dialogClasses.root, className, classes?.root)}
          id={id}
          ref={measuredRef}
          open={open}
          fullScreen={fullscreen}
          onClose={(event, reason) => wrappedClose(event, undefined, reason)}
          slots={slots}
          classes={{ container: css({ position: "relative" }) }}
          BackdropProps={{
            classes: {
              root: cx(dialogClasses.background, classes?.background),
            },
          }}
          PaperProps={{
            classes: {
              root: cx(
                dialogClasses.paper,
                fullscreen && cx(dialogClasses.fullscreen, "fullscreen"),
                css(styles.paper),
                css({ position: "absolute" }),
                classes?.paper,
                fullscreen && classes?.fullscreen
              ),
            },
          }}
          aria-modal
          {...others}
        >
          <StyledClose
            id={setId(id, "close")}
            className={cx(dialogClasses.closeButton, classes?.closeButton)}
            variant="secondaryGhost"
            onClick={(event) => wrappedClose(event, true, undefined)}
            aria-label={buttonTitle}
          >
            <CloseButtonTooltipWrapper />
          </StyledClose>
          {children && typeof children === "object"
            ? React.Children.map(
                children,
                (c: React.ReactNode) =>
                  c &&
                  React.cloneElement(c as React.ReactElement, { fullscreen })
              )
            : children}
        </MuiDialog>
      )}
    </ClassNames>
  );
};
