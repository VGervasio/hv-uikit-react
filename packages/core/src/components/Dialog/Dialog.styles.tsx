import styled from "@emotion/styled";
import Backdrop from "@mui/material/Backdrop";
import Dialog from "@mui/material/Dialog";
import Paper from "@mui/material/Paper";
import { theme } from "@hitachivantara/uikit-styles";
import { transientOptions } from "utils/transientOptions";
import fade from "utils/hexToRgbA";
import { HvButton } from "components";

export const StyledDialog = styled(Dialog)({});

export const StyledPaper = styled(
  Paper,
  transientOptions
)(({ $fullscreen }: { $fullscreen: boolean }) => ({
  color: theme.colors.acce1,
  flex: $fullscreen ? "1" : "0 0 66.66667%",
  maxWidth: $fullscreen ? "100%" : "66.66667%",
  backgroundColor: theme.colors.atmo1,
  padding: "0px",
  overflow: "auto",
  boxShadow: ["none", "0 2px 12px rgba(65,65,65,0.12)"],
  borderRadius: theme.dialog.borderRadius,
  "&:not(.fullscreen)": {
    maxHeight: `calc(100% - (2 * ${theme.dialog.margin}))`,
    display: "flex",
    flexDirection: "column",
  },
  "& svg *.color0": {
    fill: theme.colors.acce1,
  },
}));

export const StyledBackdrop = styled(
  Backdrop,
  transientOptions
)(({ $backColor }: { $backColor: string }) => ({
  background: fade($backColor, 0.8),
}));

export const StyledClose = styled(HvButton)({
  padding: 0,
  minWidth: "inherit",
  position: "absolute",
  top: theme.spacing(2),
  right: theme.spacing(2),
});
