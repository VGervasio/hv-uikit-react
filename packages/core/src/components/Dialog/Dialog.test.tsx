import { render, fireEvent } from "@testing-library/react";
import { describe, expect, it, vi } from "vitest";
import { HvDialog } from "./Dialog";
import { HvDialogActions, HvDialogContent, HvDialogTitle } from "./index";

describe("Dialog", () => {
  it("should be defined", () => {
    const { container } = render(<HvDialog open />);
    expect(container).toBeDefined();
  });

  it("should render correctly", () => {
    const { container } = render(<HvDialog open />);
    expect(container).toMatchSnapshot();
  });

  it("should render all components correctly", () => {
    const { container, getByText } = render(
      <HvDialog open>
        <HvDialogTitle>mockTitle</HvDialogTitle>
        <HvDialogContent>mockContent</HvDialogContent>
        <HvDialogActions>mockActions</HvDialogActions>
      </HvDialog>
    );
    expect(getByText("mockTitle")).toBeInTheDocument();
    expect(getByText("mockContent")).toBeInTheDocument();
    expect(getByText("mockActions")).toBeInTheDocument();
    expect(container).toMatchSnapshot();
  });

  it("should call the onClose function when the close button is clicked", () => {
    const mockFn = vi.fn();
    const mockBtnTitle = "closeBtn";

    const { getByRole } = render(
      <HvDialog open onClose={mockFn} buttonTitle={mockBtnTitle} />
    );
    const btn = getByRole("button", { name: mockBtnTitle });
    fireEvent.click(btn);
    expect(mockFn).toHaveBeenCalled();
  });
});
