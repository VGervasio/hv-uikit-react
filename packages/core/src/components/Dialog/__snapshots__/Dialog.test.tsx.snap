// Vitest Snapshot v1

exports[`Dialog > should render all components correctly 1`] = `
<div
  aria-hidden="true"
/>
`;

exports[`Dialog > should render correctly 1`] = `
<div
  aria-hidden="true"
/>
`;
